src/librustc/ty/relate.rs

```diff
@@ -430,12 +432,40 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
             let t = relation.relate(&a_t, &b_t)?;
             assert_eq!(sz_a.ty, tcx.types.usize);
             assert_eq!(sz_b.ty, tcx.types.usize);
-            let sz_a_u64 = sz_a.val.to_const_int().unwrap().to_u64().unwrap();
-            let sz_b_u64 = sz_b.val.to_const_int().unwrap().to_u64().unwrap();
-            if sz_a_u64 == sz_b_u64 {
-                Ok(tcx.mk_ty(ty::TyArray(t, sz_a)))
-            } else {
-                Err(TypeError::FixedArraySize(expected_found(relation, &sz_a_u64, &sz_b_u64)))
+            let to_u64 = |x: &'tcx ty::Const<'tcx>| -> Option<u64> {
+                match x.val {
+                    ConstVal::Integral(x) => Some(x.to_u64().unwrap()),
+                    ConstVal::Unevaluated(def_id, substs) => {
+                        // FIXME(eddyb) get the right param_env.
+                        let param_env = ty::ParamEnv::empty(Reveal::UserFacing);
+                        match tcx.lift_to_global(&substs) {
+                            Some(substs) => {
+                                match tcx.const_eval(param_env.and((def_id, substs))) {
+                                    Ok(&ty::Const { val: ConstVal::Integral(x), .. }) => {
+                                        return Some(x.to_u64().unwrap());
+                                    }
+                                    _ => {}
+                                }
+                            }
+                            None => {}
+                        }
+                        tcx.sess.delay_span_bug(tcx.def_span(def_id),
+                            "array length could not be evaluated");
+                        None
+                    }
+                    _ => bug!("arrays should not have {:?} as length", x)
+                }
+            };
+            match (to_u64(sz_a), to_u64(sz_b)) {
+                (Some(sz_a_u64), Some(sz_b_u64)) => {
+                    if sz_a_u64 == sz_b_u64 {
+                        Ok(tcx.mk_ty(ty::TyArray(t, sz_a)))
+                    } else {
+                        Err(TypeError::FixedArraySize(
+                            expected_found(relation, &sz_a_u64, &sz_b_u64)))
+                    }
+                }
+                _ => Ok(tcx.types.err)
```

Comment:
So, it's ok to use `err` here because we currently expect these to always be evaluatable? Can we add a comment about it? Also, I'd prefer the return type of the closure to be `Result<u64, ErrorRepored>`, rather than `Option<u64>`, since it makes it clearer in the type what is expected, and we are less likely to make the closure return `None` without thinking about it (I personally would also write `(Err(ErrorReported), _) | (_, Err(ErrorReported)) => {` instead of `_`, as insurance against the type changing...)

Reply:
I have a `delay_span_bug` so it's impossible to produce `tcx.types.err` without a compiler error. We need lazy normalization to really do anything smarter here.